A69-crl-enhancements.md

```diff
@@ -0,0 +1,226 @@
+1. A `CrlReloaderProvider` that periodically and asynchronously reloads a local directory of CRLs.
+
+Users can also implement the interface with whatever behaviors they specifically desire.
+
+Following [RFC5280](https://datatracker.ietf.org/doc/html/rfc5280), during revocation checking one of three status should be returned - `RevocationUnrevoked`, `RevocationRevoked`, and `RevocationUndetermined` ([RFC5280 defines `Undetermined` in the CRL Processing section](https://datatracker.ietf.org/doc/html/rfc5280#section-6.3.3). We ). There are many reasons a certificate can be revoked, but in the end gRPC cares about a certificate being revoked or not, thus we differentiate between `RevocationUnrevoked` and `RevocationRevoked`. The user must specify whether `RevocationUndetermined` should be treated as `RevocationRevoked` or `RevocationUnrevoked` (failing open vs. failing closed).
+
+Assumptions:
+1. There should be no direct linkage between this interface and the creation/distribution of CRLs. gRPC is a user of CRLs and credentials, not a PKI itself.
+1. The CRL Providers simply provide CRLs. The actual verification will be done by other pieces of the code (these are already written).
+1. We cannot know users' PKI details, so our implementations should not enforce non-X509 metadata mattering (for example, file names). A user could certainly implement their own provider that cares about filenames.
+1. We will not support OSCP-style checking (OSCP is an alternative to CRLs, not another form of CRLs).
+
+![Basic Flow Diagram](A69_graphics/basic_diagram.png)
+
+CrlProvider Interface Pseudocode:
+```
+public:
+// Get the CRL that is associated with a cert. Provides good semantic
+// consistency as most X509 APIs that try to get a CRL expect to get a cert
+// passed in. In each language we may use an abstraction here that makes it easy
+// to get CRL-lookup information from the certificate.
+// Return value expected to be as read-only 
+// Returns null if there isn't an associated CRL (as null is different than an
+// empty CRL)
+  CRL Crl(certficate/certificateInfo)
+private:
+  // Implementation for getting and storing CRLs
+```
+
+Example feature use:
+```
+opts = TlsOptions()
+provider = CrlDirectoryReloaderProvider("/path/to/crl/directory", Duration(60 seconds))
+opts.provider = provider
+server = Server(opts)
+```
+
+### Map Content and API Outcomes
+
+The diagram below describes the behavior of the API and why this or that outcome is produced. Here, "Map" is a generic way to represent some in-memory representation of the CRLs in a key-value store. “CRL Provider API Return” refers to the CRL return of the `CrlProvider` in the Pseudocode above.
+![Basic API Outcomes](A69_graphics/basic_table.png)
+
+The pseudo code snippet below illustrates the same idea:
+```
+<pseudo code in the handshaker when verifying cert>
+
+RevocationStatus GetRevocationStatus(cert, crlProvider, settings) {
```

Comment:
Suggest changing this pseudo-code as follows:
- Rename `settings` to `opts`, to clarify that this is the same options structure created in the previous pseudo-code block.
- Instead of passing in `crlProvider` as a separate parameter, use `opts.provider`, since that's where we set the provider in the previous pseudo-code block.